src/util/codec/mysql/json.rs

```diff
@@ -0,0 +1,561 @@
+        let mut encode_values = vec![];
+        for value in data.values() {
+            try!(value_entries.encode_json_item(value, &mut value_offset, &mut encode_values));
+        }
+        let size = value_offset;
+        try!(self.encode_u32_le(element_count as u32));
+        try!(self.encode_u32_le(size as u32));
+        try!(self.write_all(key_entries.as_mut()));
+        try!(self.write_all(value_entries.as_mut()));
+        try!(self.write_all(encode_keys.as_mut()));
+        try!(self.write_all(encode_values.as_mut()));
+        Ok(())
+    }
+
+    fn encode_array(&mut self, data: &[Json]) -> Result<()> {
+        // array ::= element-count size value-entry* value*
+        let element_count = data.len();
+        let count_len = LENGTH_U32;
+        let size_len = LENGTH_U32;
+        let value_entries_len = LENGTH_VALUE_ENTRY * element_count;
+        let mut value_offset = (count_len + size_len + value_entries_len) as u32;
+        let mut value_entries = Vec::with_capacity(value_entries_len);
+        let mut encode_values = vec![];
+        for value in data {
+            try!(value_entries.encode_json_item(value, &mut value_offset, &mut encode_values));
+        }
+        let total_size = value_offset;
+        try!(self.encode_u32_le(element_count as u32));
+        try!(self.encode_u32_le(total_size as u32));
+        try!(self.write_all(value_entries.as_mut()));
+        try!(self.write_all(encode_values.as_mut()));
+        Ok(())
+    }
+
+    fn encode_literal(&mut self, data: u8) -> Result<()> {
+        try!(self.write_u8(data));
+        Ok(())
+    }
+
+    fn encode_json_i64(&mut self, data: i64) -> Result<()> {
+        self.encode_i64_le(data)
+    }
+
+    fn encode_json_f64(&mut self, data: f64) -> Result<()> {
+        self.encode_f64_le(data)
+    }
+
+    fn encode_str(&mut self, data: &str) -> Result<()> {
+        let bytes = data.as_bytes();
+        let bytes_len = bytes.len() as u64;
+        try!(self.encode_var_u64(bytes_len));
+        try!(self.write_all(bytes));
+        Ok(())
+    }
+
+    fn encode_json_item(&mut self,
+                        data: &Json,
+                        offset: &mut u32,
+                        data_buf: &mut Vec<u8>)
+                        -> Result<()> {
+        let code = data.get_type_code();
+        try!(self.write_u8(code));
+        match *data {
+            // If the data has length in (0, 4], it could be inline here.
+            // And padding 0x00 to 4 bytes if needed.
+            Json::Literal(ref v) => {
+                try!(self.write_u8(*v));
+                let left = LENGTH_U32 - LENGTH_LITERAL;
+                for _ in 0..left {
+                    try!(self.write_u8(JSON_LITERAL_NIL));
+                }
+            }
+            _ => {
+                try!(self.encode_u32_le(*offset));
+                try!(data_buf.encode_json_body(data));
+                *offset += (data.binary_len() - LENGTH_TYPE) as u32;
```

Comment:
I think `data.binary_len() - LENGTH_TYPE` should be equal to `length of data_buf - original length of data_buf`.